meld/cluster_topdown.py

```diff
@@ -0,0 +1,305 @@
+                 - np.power(lower_limit, h_plus1)) / h_plus1
+
+clus_spec = [
+    ()
+]
+
+
+class Cluster():
+    """ TFCE cluster class
+    """
+    def __init__(self, param_e, param_h):
+        """
+            Initialize empty cluster
+        """
+        self.param_e = param_e
+        self.h_plus1 = param_h + 1
+        # summed TFCE over all cluster members
+        self.accum_val = 0.
+        # extent of cluster; called "totalArea" in Cpp implementation
+        self.extent = 0.
+        # list of cluster members
+        self.members = []
+        # upper limit of TFCE integral; called "lastVal" in Cpp implementation
+        self.upper_limit = 0.
+
+    def update(self, lower_limit):
+        """
+            Perform an incremental update to the summed TFCE integral over
+            cluster members
+
+            Args
+            ----
+            lower_limit : float
+                lower limit of TFCE integral
+                called "bottomVal" in Cpp implementation
+        """
+        if len(self.members) == 0:
+            self.upper_limit = lower_limit
+        elif not lower_limit == self.upper_limit:
+            assert lower_limit < self.upper_limit
+            new_slice_val = tfce_calc(self.extent,
+                                      self.param_e,
+                                      self.upper_limit,
+                                      self.h_plus1,
+                                      lower_limit)
+            self.accum_val += new_slice_val
+            self.upper_limit = lower_limit
+
+    def add_member(self, node, height, extent):
+        """
+            Add a new node to cluster
+
+            Args
+            ----
+            node : int
+                Node index
+            height : float
+                Test statistic value at node
+                called "val" in Cpp implementation
+            extent : float
+                Extent of node
+                called "area" in Cpp implementation
+        """
+        self.update(height)
+        # append node to list of members
+        self.members.append(node)
+        self.extent += extent
+
+
+def alloc_cluster(cluster_list, dead_clusters, param_e, param_h):
+    """
+        Generate a valid (new) cluster number
+        If available, number for a dead cluster will be recycled
+        Cluster and dead cluster lists will be updated
+
+        Args
+        ----
+            cluster_list : list of Clusters
+            dead_clusters : set of Clusters
+            param_e : float
+            param_h : float
+    """
+    # verify that the lists are actually being modified in main code
+    if len(dead_clusters) == 0:
+        cluster_list.append(Cluster(param_e, param_h))
+        return len(cluster_list) - 1
+    else:
+        # removes and returns element from set
+        ret = dead_clusters.pop()
+        cluster_list[ret] = Cluster(param_e, param_h)
+        return ret
+
+
+def get_node_neighbors(node, connectivity):
```

Comment:
With how simple this function is (now, anyway), you might consider removing it and putting the equivalent in the code itself, in case the python runtime is making an extra copy just because it is in a function.

Reply:
Mutable objects like lists and dictionaries are passed to functions as references in Python, so at worst it would have to copy the node integer, but it's still an easy change to make.